core/kernel/ubsan.c

```diff
@@ -75,36 +75,25 @@ struct invalid_builtin_data {
  */
 void __ubsan_handle_type_mismatch(struct type_mismatch_data *data,
 				  unsigned long ptr);
-void __ubsan_handle_type_mismatch_v1(struct type_mismatch_data *data,
-				     unsigned long ptr);
-void __ubsan_handle_add_overflow(struct overflow_data *data,
-				  unsigned long lhs, unsigned long rhs);
-void __ubsan_handle_sub_overflow(struct overflow_data *data,
-				  unsigned long lhs, unsigned long rhs);
-void __ubsan_handle_mul_overflow(struct overflow_data *data,
-				  unsigned long lhs, unsigned long rhs);
-void __ubsan_handle_negate_overflow(struct overflow_data *data,
-				    unsigned long old_val);
-void __ubsan_handle_divrem_overflow(struct overflow_data *data,
-				    unsigned long lhs, unsigned long rhs);
-void __ubsan_handle_pointer_overflow(struct overflow_data *data,
-				     unsigned long lhs, unsigned long rhs);
-void __ubsan_handle_shift_out_of_bounds(struct shift_out_of_bounds_data *data,
-					unsigned long lhs, unsigned long rhs);
-void __ubsan_handle_out_of_bounds(struct out_of_bounds_data *data,
-				  unsigned long idx);
+void __ubsan_handle_type_mismatch_v1(void *data_, void *ptr);
```

Comment:
How did you find the proper declaration of this and the other `__ubsan_handle_` functions?
Do you have a link?

Reply:
Hi,

There are at least 3 ways

1) gcc gives warning during compilation and shows correct prototype at the same time. So one can rely on this gcc warning messages to figure out correct function signature. The initial idea behind this was to fix warnings.
For example:
```
core/kernel/ubsan.c:78:6: warning: conflicting types for built-in function ‘__ubsan_handle_type_mismatch_v1’; expected ‘void(void *, void *)’ [-Wbuiltin-declaration-mismatch]
   78 | void __ubsan_handle_type_mismatch_v1(void *data_, unsigned long ptr);
```
2) one can refer to [linux kernel implementation](https://github.com/torvalds/linux/blob/master/lib/ubsan.c#L402)
3) [look](https://github.com/gcc-mirror/gcc/blob/master/gcc/sanitizer.def#L515) into gcc . I hope I gave a link to correct place